vcf/toVariant.go

```diff
@@ -0,0 +1,136 @@
+package vcf
+
+import (
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/variant"
+	"log"
+	"strings"
+)
+
+// SmallVariantChans wraps channels for all of the small variants that
+// are valid variant.Mutators and variant.Effectors. All channels except
+// for Records sends a parsed version of a given Vcf record that stores
+// the corresponding variant type. Each send on one of the variant channels
+// must be paired with a send on the records channel. i.e. for each parsed
+// variant sent, the original vcf record is the following send. This is
+// useful as the Vcf records often hold metadata that is not stored in the
+// variant struct. Even if the original Vcf record is not used, the Records
+// channel must be read else the sending goroutine will be blocked.
+type SmallVariantChans struct {
+	Substitutions chan variant.Substitution
+	Insertions    chan variant.Insertion
+	Deletions     chan variant.Deletion
+	Delins        chan variant.Delins
+	Records       chan Vcf
+}
+
+// NewSmallVariantChans creates a series of channels to send the variant
+// representation of a given vcf record. The input buffSize determines
+// the channel buffer size for all on the variant channels.
+func NewSmallVariantChans(buffSize int) SmallVariantChans {
+	var answer SmallVariantChans
+	answer.Substitutions = make(chan variant.Substitution, buffSize)
+	answer.Insertions = make(chan variant.Insertion, buffSize)
+	answer.Deletions = make(chan variant.Deletion, buffSize)
+	answer.Delins = make(chan variant.Delins, buffSize)
+	answer.Records = make(chan Vcf, buffSize)
+	return answer
+}
+
+// ChanToVariants splits an incoming channel of VCFs to a channel of variant types.
+// Must be run as a goroutine else the thread will deadlock.
+// See documentation for SmallVariantChans for more information on variant channels.
+func ChanToVariants(c <-chan Vcf, sendChans SmallVariantChans) {
+	for v := range c {
+		if !canParseToBases(v.Ref) {
+			continue // cannot parse vcf record so skip
+		}
+		var refSeq, altSeq []dna.Base
```

Comment:
Would it be better to put these up outside the for loop so that we don't keep re-declaring them?

Reply:
Yes, good catch